### DISCUSSION TITLE
React에서 빌트인 생성자 함수는 초기 렌더링이 완료되고 window 객체가 생성된 이후에 사용할 수 있나요?

### DISCUSSION THREAD

**Discussion by drizzle96** [codeit-bootcamp-frontend/0-study-js-deepdive#7] [Q&A]
빌트인 생성자 함수는 전역 객체(브라우저의 경우 window)가 생성되는 시점에 생성된다고 합니다. (참고: 19.5.2절)

그리고 React에서 window 객체는 초기 렌더링이 완료된 후 정의되는 것으로 알고 있습니다.
[참고: 팀 데일리 미션 - Q. localStorage는 어떤 객체의 프로퍼티이며, 리액트에서 로컬스토리지를 사용할 때 주의점을 알려주세요]( https://www.notion.so/Q-localStorage-b5a0c2d095f34458a94cdf8786267dc1)

따라서 빌트인 생성자 함수 역시 초기 렌더링이 완료되기 전에는 사용할 수 없을지 궁금합니다.

**Comment by SeyoungCho** [comment]
@drizzle96  페어프로그래밍 때 잘못된 개념을 알려드렸네요. window 객체는 브라우저에서 코드가 실행되기 이전에 생성된다고 합니다. 그래서 리액트 어디서든 사용할 수 있는게 맞고요. 당연히 로컬스토리지도 사용가능합니다. useState initial 값으로도 사용할 수 있더라고요. window객체도 마찬가지입니다. 제가 잘못알고 있었던 것은 예전에 프로젝트 할 때 Next.js로 서버사이드 코드 개발을 할 때 localStorage를 접근하다가 서버 사이드에는 브라우저 환경이 아니라서 해당 코드가 클라이언트로 넘어간 이후에야 window객체를 참조할 수 있었습니다. 그래서 if (typeof window !== 'undefined')라는 조건문이 나온거고요. 
죄송합니다..!! 다음엔 확실하지 않은건 확실히 찾아보고 말씀드리겠습니다!! 질문 감사해요 ~덕분에 오개념 바로잡았습니다.

\=======================
이전 프로젝트 때 기록해뒀던 내용입니다
\=======================

- Localstorage is not defined
    
    로컬 스토리지는 원래 window객체의 프로퍼티객체이므로 window 객체가 아직 정의되지 않은 시간동안은 localStorage에 접근이 불가능하고 에러가 뜬다. 따라서 window의 타입을 narrowing해주는 블록 안에서 localStorage에 접근할 수 있다. (특히 마운팅과 동시에 localstorage에 접근하는 경우 이런 에러가 많이 뜬다)
    
    ```jsx
    if(typeof window !== 'undefined'){
    	localStorage.getItem(...);
    }
    ```

**Comment by drizzle96** [comment]
아하 그런거였군요! 죄송할 필요 없고 저도 감사합니다~!! 어제 구글링 하다가 아래 글 보고 저도 window 객체가 렌더링이 된 후 정의가 되나보나 했었는데 여기서도 next.js 때 얘기였네요 

https://developer.school/snippets/react/localstorage-is-not-defined-nextjs

**Comment by drizzle96** [comment]
해결
